### PR TITLE
#2936  Issue on [cli/info] multipass info --all does not list deleted instances : modify the list_instances function in the multipass-info.cpp and also…

### PR DESCRIPTION
**Pull request by Aryansingh0103**
… added main  function to pass the show_deleted flag to it.

I've added a show_deleted parameter to control whether information about deleted instances should be included in the output. I've also added a check for deleted instances and a loop to print information about each deleted instance if the show_deleted flag is set. And In order to use the modified list_instances function, we need to modify the info_main function to pass the show_deleted flag to it.

### CODE REVIEW

File: .vscode/c_cpp_properties.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (georgeliao) [inline .vscode/c_cpp_properties.json:1]: Why is this vscode file here? Is it something local to you and it was added accidentally?

**Review by andrei-toterman — CHANGES_REQUESTED** [review]
Hi, @Aryansingh0103! I'm glad that you've decided to try to contribute to Multipass!

First things first: I see that you've committed your `.vscode` directory. That should definitely be avoided, as it's specific only to your setup. To be honest, we could also improve our `.gitignore` file with some more common entries like that one.

Now, for the actual fix itself, let me give you a brief explanation about how Multipass functions. There are two main components: the daemon `multipassd` and the CLI `multipass`. The daemon is a process that always runs in the background and manages the state of the virtual machines. The CLI is a tool that you use in your terminal in order to interact with the daemon, by sending it requests and displaying its replies back to the user.

When you do `multipass info --all` in the terminal, you use the `multipass` CLI to send a request to the `multipassd` daemon to give you all the information about the instances. The request that is being sent is correct, but the daemon does not return a complete reply. Therefore, we have nothing to fix within the CLI, only within the daemon code.

Each request that the daemon may receive is handled by a related method inside the `Daemon` class. In the case of `multipass info`, the request is handled inside the `Daemon::info` method. Inside that method, we look at the received request and we construct a reply with the appropriate information. The important part of the code for this issue is this one, found [here](https://github.com/canonical/multipass/blob/main/src/daemon/daemon.cpp#LL1346-L1357)
```cpp
std::vector<decltype(vm_instances)::key_type> instances_for_info;

if (request->instance_names().instance_name().empty())
{
    for (auto& pair : vm_instances)
        instances_for_info.push_back(pair.first);
}
else
{
    for (const auto& name : request->instance_names().instance_name())
        instances_for_info.push_back(name);
}
```

The first vector, `instances_for_info`, is a vector of strings representing the names of the VMs that we want to query for information. In the following `if` statement, we check if the request we received has any VM names specified. If it doesn't, as in the case of `multipass info --all`, we go through our non-deleted instances, which we keep inside the `vm_instances` hashmap, and we add the name of that VM, which is used as a key in the hashmap and retrieved from `pair.first`, to the list of VM names that we're interested into. Conversely, if we do specify VM names in `multipass info`, we go on the `else` branch and we only add the names that we specified to `instances_for_info`.

There is no issue with what happens with `instances_for_info` after this point. The issue is that we keep deleted and non-deleted VMs in two different hashmaps. The non-deleted ones are, as seen, inside `vm_instances`. The deleted ones are inside a hashmap called `deleted_instances`. And when we want to add all the VMs to `instances_for_info`, we only iterate through `vm_instances`, but not through `deleted_instances` as well. So I suggest that you do something similar with
```cpp
for (auto& pair : vm_instances)
    instances_for_info.push_back(pair.first);
```
but for the `deleted_instances` as well.

I hope this explanation clears things up a bit. Don't hesitate to update the PR as many times as needed :smile:

**Comment by Aryansingh0103** [comment]
can I add this code 
std::vector<decltype(vm_instances)::key_type> instances_for_info;

for (auto& pair : vm_instances)
{
    instances_for_info.push_back(pair.first);
}

for (auto& pair : deleted_instances)
{
    auto it = std::find(instances_for_info.begin(), instances_for_info.end(), pair.first);
    if (it == instances_for_info.end())
    {
        instances_for_info.push_back(pair.first);
    }
}

With these changes, the instances_for_info vector will contain the names of all non-deleted and deleted VMs, ensuring that all VMs are included in the output of the multipass info --all command. The code also checks for duplicates before adding VM names to the instances_for_info vector to avoid including the same VM multiple times.

File: .vscode/c_cpp_properties.json (hunk 1 above)
Comment (Aryansingh0103) [reply .vscode/c_cpp_properties.json:1]: it was added accidentally

**Comment by andrei-toterman** [comment]
You're on the right track! It's true that we don't want duplicates in the `instances_for_info` vector, but since `vm_instances` holds all non-deleted instances and `deleted_instances` holds all the deleted ones, there is no overlap between them, so you don't need to manually check for duplicates, as they simply won't happen.

**Comment by andrei-toterman** [comment]
Hi, @Aryansingh0103! Unfortunately I must close this PR, as we need a fix for this very soon. I hope you'll find other opportunities to contribute to Multipass or other open source projects :smile:

**Pull request closed without merging**